File: 2026/07/31/closures/haskell-dispenser/src/Dispenser.hs
module Dispenser (makeDispenser, findFirst, findNext)
    where

import Data.List (findIndex, (!!))

type Item = Int
type Index = Int
type Result = Maybe (Index, Item)
type Predicate = Item -> Bool

index :: (Index, Item) -> Index
index = fst

item :: (Index, Item) -> Item
item = snd

data Dispenser = Dispenser
    {
    items :: [Item],
    predicate :: Maybe Predicate,
    position :: Maybe Index
    }

makeDispenser :: [Item] -> Dispenser
makeDispenser items = Dispenser
    {
    items = items,
    predicate = Nothing,
    position = Nothing
    }

-- set the new predicate and find the first matching item
findFirst :: Dispenser -> Predicate -> (Maybe Item, Dispenser)
findFirst dispenser pred = findNext newDispenser
    where
        newDispenser = dispenser
            {
            predicate = Just pred,
            position = Just 0
            }

-- find the next matching item, set the new dispenser
findNext :: Dispenser -> (Maybe Item, Dispenser)
findNext (Dispenser _ Nothing _) 
  = error "no predicate in this dispenser; call findFirst first"

findNext dispenser@(Dispenser _ _ Nothing) = (Nothing, dispenser)

findNext dispenser@(Dispenser items (Just predicate) (Just pos)) =
    (item <$> result, dispenser { position = nextPosition } )
        where
            nextPosition =  (\idx -> pos + idx + 1) <$> index <$> result 
            result = findItem (drop pos items) predicate

findItem :: [Item] -> Predicate -> Result
findItem list aPredicate = findIndex aPredicate list >>= \pos -> Just (pos, list !! pos)

File: 2026/07/31/closures/rust-dispenser/src/dispenser.rs
use std::sync::Arc;

type Item = u16;

#[derive(Clone)]
pub struct Predicate {
    pub function: Arc<dyn Fn(Item) -> bool>,
}

#[allow(dead_code)]
impl Predicate {
    pub fn new<T: Fn(Item) -> bool + 'static>(function: T) -> Self {
        Predicate {
            function: Arc::new(function),
        }
    }
}

#[allow(dead_code)]
pub struct Dispenser {
    items: Vec<Item>,
    predicate: Option<Predicate>,
    position: Option<usize>,
}

#[allow(dead_code)]
impl Dispenser {
    pub fn new(items: Vec<Item>) -> Self {
        Dispenser {
            items: items.clone(),
            predicate: None,
            position: None,
        }
    }

#[allow(dead_code)]
    pub fn find_first(&mut self, predicate: Predicate) -> Option<Item> {
        self.predicate = Some(predicate.clone());
        if let Some((next_position, value))= find_item(&self.items, &predicate, 0) {
            self.position = Some(next_position);
            Some(value)
        } else {
            self.position = None;
            None
        }
    }

    pub fn find_next(&mut self) -> Option<Item> {
        let predicate = match &self.predicate {
            Some(predicate) => predicate,
            None => panic!("no predicate; call find_first first"),
        };
        let position = match self.position {
            Some(position) => position,
            None => return None,
        };
        if let Some((next_position, value)) = find_item(&self.items, predicate, position) {
            self.position = Some(next_position);
            Some(value)
        } else {
            self.position = None;
            None
        }
    }
}

#[allow(dead_code)]
fn find_item(items: &Vec<Item>, predicate: &Predicate, position: usize) -> Option<(usize, Item)> {
    let function = predicate.function.clone();
    let start_position = position;
    let index = items[position..]
        .iter()
        .position(|item| function(*item))
        .map(|i| position + i)?;
    let position = index+1;
    if position == start_position {
        None
    } else {
        Some((position, items[index]))
    }
}
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_dispenser_can_find_the_items_matching_a_predicate() {
        let mut dispenser = Dispenser::new(vec![23, 17, 42, 4807, 3, 12, 256, 100, 1017 ]);
        let predicate = Predicate::new(|n| n >= 100);
        let result = dispenser.find_first(predicate);
        assert_eq!(result, Some(4807));
        let result = dispenser.find_next();
        assert_eq!(result, Some(256));
        let result = dispenser.find_next();
        assert_eq!(result, Some(100));
        let result = dispenser.find_next();
        assert_eq!(result, Some(1017));
        let result = dispenser.find_next();
        assert_eq!(result, None);
    }
}
